From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BayLee4 <63376748+BayLee4@users.noreply.github.com>
Date: Fri, 23 Oct 2026 00:28:56 +0200
Subject: [PATCH] 09992: Fix call-site overrides returning a struct through a
 hidden pointer

9991: Fix call-site overrides returning a struct through a
 hidden pointer
---
 Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc     | 3 ++-
 .../ghidra/program/model/pcode/HighFunctionDBUtil.java    | 8 +++++++-
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
index c39c38ce56..192b35947f 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
@@ -3762,6 +3762,7 @@ void ProtoStoreInternal::decode(Decoder &decoder,ProtoModel *model)
   }
   decoder.closeElement(elemId);
   ProtoParameter *curparam;
+  bool outputLocked = (pieces[0].flags & ParameterPieces::typelock)!=0;
   if (!addressesdetermined) {
     // If addresses for parameters are not provided, use
     // the model to derive them from type info
@@ -3788,7 +3789,7 @@ void ProtoStoreInternal::decode(Decoder &decoder,ProtoModel *model)
   for(uint4 i=1;i<pieces.size();++i) {
     if ((pieces[i].flags&ParameterPieces::hiddenretparm)!=0) {
        curparam = setInput(i-1,"rethidden",pieces[i]);
-       curparam->setTypeLock((pieces[0].flags & ParameterPieces::typelock)!=0);   // Has output's typelock
+       curparam->setTypeLock(outputLocked);   // Has output's typelock
        continue;    // increment i but not j
     }
     curparam = setInput(i-1,proto.innames[j],pieces[i]);
diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/HighFunctionDBUtil.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/HighFunctionDBUtil.java
index 8233eff8e7..6a248980dc 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/HighFunctionDBUtil.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/pcode/HighFunctionDBUtil.java
@@ -815,7 +815,13 @@ public class HighFunctionDBUtil {
 				Address callDestAddr = ref.getToAddress();
 				Function func = program.getFunctionManager().getFunctionAt(callDestAddr);
 				if (func != null && func.hasVarArgs()) {
-					return func.getParameterCount();
+					int count = 0;
+					for (Parameter param : func.getParameters()) {
+						if (param.getAutoParameterType() != AutoParameterType.RETURN_STORAGE_PTR) {
+							count++;
+						}
+					}
+					return count;
 				}
 				break;
 			}
-- 
2.45.1

